// wdio-teamcity-reporter/index.js
'use strict';

var _require = require('events'),
    EventEmitter = _require.EventEmitter;

var _require2 = require('./lib/message'),
    buildFormatter = _require2.buildFormatter,
    events = _require2.events;

var _require3 = require('lodash'),
    flow = _require3.flow;

var _require4 = require('util'),
    inherits = _require4.inherits;

module.exports = WdioTeamcityReporter;
module.exports.reporterName = 'teamcity';

/**
 * @param {object}  baseReporter
 * @param {object}  wdioConf
 * @param {object}  reporterOptions
 * @param {boolean} reporterOptions.captureStandardOutput
 * @param {boolean} reporterOptions.flowId
 * @param {string}  reporterOptions.message
 * @return {wdioTeamcityReporter}
 */
function WdioTeamcityReporter(baseReporter, wdioConf) {
  var reporterOptions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  this.baseReporter = baseReporter;

  var opts = {
    captureStandardOutput: typeof reporterOptions.captureStandardOutput === 'boolean' ? reporterOptions.captureStandardOutput : false,
    flowId: typeof reporterOptions.flowId === 'boolean' ? reporterOptions.flowId : true,
    name: typeof reporterOptions.message === 'string' ? reporterOptions.message : '[title]'
  };

  this.enableRealTimeOutput(opts);
}

inherits(WdioTeamcityReporter, EventEmitter);

WdioTeamcityReporter.prototype.enableRealTimeOutput = function (opts) {
  var _this = this;

  events.forEach(function (event) {
    return _this.on(event, flow(buildFormatter(event, opts), output));
  });
};

/**
 * @param  {string} msg
 * @return {string}
 */
function output(msg) {
  if (msg) {
    console.log(msg);
  }
}
